✨ New: Snowflake Cortex Destination 🚀

bindipankhudi opened the pull request:
Related to: https://github.com/airbytehq/PyAirbyte/issues/120

Destination is working as expected. All unit tests and integration tests are passing. Majority of the code is generated from the `vector-db SDK`, or taken from Pinecone destination. The most interesting bit (write logic), specific to Cortex is in `SnowflakeCortexIndexer` class.

To be done as fast follow
* Add an option for embedding via Snowflake models.

bindipankhudi commented on airbyte-integrations/connectors/destination-snowflake-cortex/destination_snowflake_cortex/indexer.py:
```diff
@@ -0,0 +1,107 @@
+        self.warehouse = config.warehouse
+        self.role = config.role
+        cache = SnowflakeCache(account=self.account, username=self.username, password=self.password, database=self.database, warehouse=self.warehouse, role=self.role)
+        # self.processor = SnowflakeSqlProcessor(cache=cache)
```
this is uncommented right not i am unable to mock them in unit tests.

bindipankhudi commented on airbyte-integrations/connectors/destination-snowflake-cortex/destination_snowflake_cortex/indexer.py:
```diff
@@ -0,0 +1,162 @@
+        )-> Iterable[AirbyteMessage]:
+        """Creates Airbyte messages from chunk records."""
+        airbyte_messages = []
+        for i in range(len(document_chunks)):
```
we can just say `for chunk in document_chunks`

bindipankhudi replied:
done.

bindipankhudi commented on airbyte-integrations/connectors/destination-snowflake-cortex/destination_snowflake_cortex/indexer.py:
```diff
@@ -0,0 +1,162 @@
+            message = AirbyteMessage(type=Type.RECORD, record=chunk.record)
+            new_data = {}
+            new_data[DOCUMENT_ID_COLUMN] = self._create_document_id(message)
+            new_data[CHUNK_ID_COLUMN] = str(uuid.uuid4().int)
```
do we have a sort order for chunks ? can use int in that case.

bindipankhudi commented on airbyte-integrations/connectors/destination-snowflake-cortex/metadata.yaml:
```diff
@@ -0,0 +1,46 @@
+    oss:
+      enabled: true
+  releaseDate: 2024-04-30 # TODO: Confirm the release date
+  releaseStage: alpha
```
could do certified and beta.

bindipankhudi replied:
done.

bindipankhudi commented on airbyte-integrations/connectors/destination-snowflake-cortex/metadata.yaml:
```diff
@@ -0,0 +1,45 @@
+  connectorSubtype: vectorstore
+  connectorType: destination
+  definitionId: d9e5418d-f0f4-4d19-a8b1-5630543638e2
+  dockerImageTag: 0.0.1
```
should this be 0.1.0?

bindipankhudi replied:
updated to 0.1.0

bindipankhudi commented on airbyte-integrations/connectors/destination-snowflake-cortex/metadata.yaml:
```diff
@@ -0,0 +1,45 @@
+        resourceRequirements:
+          memory_limit: 2Gi
+          memory_request: 2Gi
+  supportLevel: certified
```
fyi - we are setting it to certified.

bindipankhudi commented on docs/integrations/destinations/snowflake-cortex.md:
```diff
@@ -0,0 +1,80 @@
+### Indexing/Data Storage 
+
+To get started, sign up for [Snowflake](https://www.snowflake.com/en/). Ensure you have set a database, and a data wareshouse before running the Snowflake Cortex destination. All streams will be indexed/stored into a table with the same name. In addition. 
+TODO: add what columns are stored and limitations on sizes 
```
add before merge

bindipankhudi replied:
done.

aaronsteers commented on airbyte-integrations/connectors/destination-snowflake-cortex/destination_snowflake_cortex/indexer.py:
```diff
@@ -0,0 +1,178 @@
+            return f"Stream_{stream_name}_Key_{primary_key}"
+        return str(uuid.uuid4().int)
+
+    def _create_state_message(self, stream, namespace, data: dict[str, Any]) -> AirbyteMessage:
```
Can we add type hints here for `stream` and `namespace` in the function signature?

bindipankhudi replied:
done.

aaronsteers commented on airbyte-integrations/connectors/destination-snowflake-cortex/destination_snowflake_cortex/indexer.py:
```diff
@@ -0,0 +1,178 @@
+                    return WriteStrategy.MERGE
+        return WriteStrategy.AUTO
+
+    def index(self, document_chunks, namespace, stream):
```
Ditto above. Can we add type hints for `document_chunks`, `namespace`, and `stream`?

bindipankhudi replied:
done.

aaronsteers commented on airbyte-integrations/connectors/destination-snowflake-cortex/destination_snowflake_cortex/indexer.py:
```diff
@@ -0,0 +1,178 @@
+            )
+            cortex_processor.process_airbyte_messages(airbyte_messages, self.get_write_strategy(stream))
+
+    def delete(self, delete_ids, namespace, stream):
```
Ditto above re: type hints. 🙂

bindipankhudi replied:
done

aaronsteers commented on airbyte-integrations/connectors/destination-snowflake-cortex/destination_snowflake_cortex/indexer.py:
```diff
@@ -0,0 +1,178 @@
+DOCUMENT_ID_COLUMN = "document_id"
+CHUNK_ID_COLUMN = "chunk_id"
+METADATA_COLUMN = "metadata"
+PAGE_CONTENT_COLUMN = "page_content"
```
Minor nit. I'm not a huge fan of "page content" as a column name, just because it seems to somewhat bias towards web paradigms.

Wdyt of something like `document_content`, `document_contents`, `text_content`, or `contents`?

bindipankhudi replied:
changed it to document_content!

aaronsteers commented on airbyte-integrations/connectors/destination-snowflake-cortex/destination_snowflake_cortex/config.py:
```diff
@@ -0,0 +1,60 @@
+
+
+# to-do - override the embedding model to use Snowflake supported model
+class SnowflakeCortexIndexingModel(BaseModel):
+    account: str = Field(
+        ...,
+        title="Account",
+        airbyte_secret=True,
+        description="Enter the account name you want to use to access the database.",
+        examples=["xxx.us-east-2.aws"],
+    )
```
I compared these with the Java-based snowflake destination to see if these parameters match the similar destination. If we can make the auth/config flow consistent between the two, this could help end users who might end up using both for different use cases.

It looks like all of our specified inputs match _except_ account, which the Java Snowflake destination calls `host` ("Host").

Also, it looks like we are not accepting `schema` ("Default Schema") and we should probably add that so users can control which schema is written to.

The last difference I noted was that the way we are configuring password is _slightly_ different, just because the Java destination supports more auth options (OAuth and Key Pair private key). Even if we just have a single auth option, it probably makes sense to try to mirror the same structure, which is to place "password" under a "credentials" object. The challenge here is that we can't necessarily copy code from the Snowflake source - since it is in Java, but maybe another Python connector that supports multiple auth flows could be a model here.

Here is the `spec.json` for `destination-snowflake`: https://github.com/airbytehq/airbyte/blob/72b083cbd259b3d922e522bee70e81d60f88f481/airbyte-integrations/connectors/destination-snowflake/src/main/resources/spec.json

bindipankhudi replied:
I will take a look at this tomm.

bindipankhudi replied:
Agreed! The parameters should ideally match the snowflake destination. Made the following changes: 
* replaced host with account. 
* reordered params to match the order in snowflake destination
* added default_schema field
* added a new credentials section with password field.
Implementing the Oauth option seems to much at the moment, but above changes should help us do so in future.

aaronsteers reviewed (changes requested):
This looks great! A few changes requested inline, but lmk if you want to push back on any of them, and/or defer.

The one that is probably hardest to implement (and also hardest to change later) is the layout of the `password` input in the config spec. Happy to help on that if needed.

bindipankhudi commented on airbyte-integrations/connectors/destination-snowflake-cortex/destination_snowflake_cortex/indexer.py:
```diff
@@ -0,0 +1,178 @@
+        """
+        self.default_processor = SnowflakeSqlProcessor(cache=self.cache)
+
+    def _get_airbyte_messsages_from_chunks(
+        self,
+        document_chunks: Iterable[Any],
+    ) -> Iterable[AirbyteMessage]:
+        """Creates Airbyte messages from chunk records."""
```
Agreed. Created a ticket to tackle this right after refactor: https://github.com/airbytehq/airbyte/issues/38172

bindipankhudi commented on airbyte-integrations/connectors/destination-snowflake-cortex/destination_snowflake_cortex/indexer.py:
```diff
@@ -0,0 +1,178 @@
+            airbyte_messages.append(message)
+        return airbyte_messages
+
+    def _get_updated_catalog(self) -> ConfiguredAirbyteCatalog:
+        """Adds following columns to catalog
+        document_id (primary key) -> unique per record/document
+        chunk_id -> unique per chunk
+        page_content -> text content of the page
+        metadata -> metadata of the record
+        embedding -> embedding of the page content
+        """
```
Agreed! Would be less error prone and much cleaner to not update catalog like this. Created ticket to tackle this after PyAirbyte refactor - https://github.com/airbytehq/airbyte/issues/38172

bindipankhudi commented on airbyte-integrations/connectors/destination-snowflake-cortex/metadata.yaml:
```diff
@@ -0,0 +1,45 @@
+data:
+  # TODO:
+  # ab_internal:
+  #   ql: 300
+  #   sl: 200
```
Will uncomment these. This is support level apparently: 😄 
`Support level, sl = (community, certified, critical) and quality level, ql = (paper, iron, bronze, silver, etc)`
I think - `sl = 200` (certified) and `ql = 300` (bronzer) make sense. Pinecone has the same values. 
For reference on the numbers: https://www.notion.so/Internal-Metadata-Fields-32b02037e7b244b7934214019d0b7cc9

bindipankhudi replied:
Done!

bindipankhudi commented on airbyte-integrations/connectors/destination-snowflake-cortex/destination_snowflake_cortex/config.py:
```diff
@@ -0,0 +1,60 @@
+from pydantic import BaseModel, Field
+
+
+# to-do - override the embedding model to use Snowflake supported model
```
Done.

bindipankhudi commented on airbyte-integrations/connectors/destination-snowflake-cortex/integration_tests/spec.json:
```diff
@@ -0,0 +1,411 @@
+            "examples": ["AIRBYTE_USER"],
+            "type": "string"
+          },
+          "credentials": {
```
Added the password in credentials section. I wanted to add a dropdown with just the "password" option, but that was not possible since dropdown with one option is not allowed, Since we have password as a second level field, we should be able to keep the same when we implement multiple authorization options.

aaronsteers replied:
This is great! And yes, makes sense! 👍

bindipankhudi commented on airbyte-integrations/connectors/destination-snowflake-cortex/integration_tests/spec.json:
```diff
@@ -0,0 +1,411 @@
+        "group": "advanced",
+        "type": "boolean"
+      },
+      "indexing": {
```
updated the order of fields to match snowflake connector.

aaronsteers replied:
Nice. 👍

bindipankhudi commented:
@aaronsteers PR is ready for another look. 
* I added all proposed minor updates (type hint, etc)
* Created this new ticket (https://github.com/airbytehq/airbyte/issues/38172) for updating how we do writes curently.
* Updated cortex config fields and order to match Snowflake destination. (more comments on this inline)

aaronsteers commented on airbyte-integrations/connectors/destination-snowflake-cortex/metadata.yaml:
```diff
@@ -0,0 +1,44 @@
+      enabled: true
+    oss:
+      enabled: true
+  releaseDate: 2024-05-13 # TODO: Confirm the release date
```
```suggestion
  releaseDate: 2024-05-14
```

aaronsteers commented on airbyte-integrations/connectors/destination-snowflake-cortex/integration_tests/sample_config.json:
```diff
@@ -0,0 +1,20 @@
+    "database": "MYDATABASE",
+    "default_schema": "MYSCHEMA",
+    "username": "MYUSERNAME",
+    "password": "xxxxxxx"
```
Should `password` be nested under `credentials` now?

Weird that this didn't throw an error.

bindipankhudi replied:
good catch.. this file is for reference purpose, doesn't get used :) i will update it.

aaronsteers commented on airbyte-integrations/connectors/destination-snowflake-cortex/destination_snowflake_cortex/config.py:
```diff
@@ -0,0 +1,88 @@
+    )
+
+    credentials: PasswordBasedAuthorizationModel
+
+    # credentials: SnowflakeCortexAuthorizationModel = Field(
+    #     default=None,
+    #     title="Authorization Method",
+    #     type="object",
+    # )
```
Should we delete this if it is no longer needed?

```suggestion
```

aaronsteers reviewed (approved):
Looks great! 🚀 

I think you can merge when ready. :shipit:
